components/collector/src/metric_collectors/test_cases.py

```diff
@@ -0,0 +1,100 @@
+"""Test cases collector."""
+
+import re
+from typing import cast, Optional, Sequence
+
+from base_collectors import MetricCollector
+from model import Entities, Entity, SourceMeasurement, MetricMeasurement
+
+
+class TestCases(MetricCollector):
+    """Test cases collector."""
+
+    __test__ = False  # Tell nose that this is not a test class
+
+    # Mapping to calculate the test result of a test case. The keys are tuples of the current test result of the test
+    # case and the next test result found. The value is the resulting test result. For example,
+    # {("passed", "skipped"), "skipped"} means that if a test case has passed so far and we see a skipped test result,
+    # the updated test result state of the test case is skipped.
+    TEST_RESULT_STATE = {
+        ("untested", "passed"): "passed",
+        ("untested", "skipped"): "skipped",
+        ("untested", "failed"): "failed",
+        ("untested", "errored"): "errored",
+        ("passed", "passed"): "passed",
+        ("passed", "skipped"): "skipped",
+        ("passed", "failed"): "failed",
+        ("passed", "errored"): "errored",
+        ("skipped", "passed"): "skipped",
+        ("skipped", "skipped"): "skipped",
+        ("skipped", "failed"): "failed",
+        ("skipped", "errored"): "errored",
+        ("failed", "passed"): "failed",
+        ("failed", "skipped"): "failed",
+        ("failed", "failed"): "failed",
+        ("failed", "errored"): "errored",
+        ("errored", "passed"): "errored",
+        ("errored", "skipped"): "errored",
+        ("errored", "failed"): "errored",
+        ("errored", "errored"): "errored",
+    }
+    TEST_CASE_KEY_RE = re.compile(r"\w+\-\d+")
+    TEST_CASE_SOURCE_TYPES = ["jira"]
+    TEST_REPORT_SOURCE_TYPES = ["junit", "testng"]
+
+    async def collect(self) -> Optional[MetricMeasurement]:
+        """Override to add the rest results from the test report(s) to the test cases."""
+        if (measurement := await super().collect()) is None:
+            return None
+        test_cases = self.test_cases(measurement.sources)
+        test_case_keys = set(test_cases.keys())
+        for entity in self.test_entities(measurement.sources):
+            for test_case_key in self.referenced_test_cases(entity) & test_case_keys:
+                test_result_so_far = test_cases[test_case_key]["test_result"]
+                test_result = entity["test_result"]
+                test_cases[test_case_key]["test_result"] = self.TEST_RESULT_STATE[(test_result_so_far, test_result)]
+        # Set the value of the test sources to zero as this metric only counts test cases
+        for test in self.test_sources(measurement.sources):
+            test.value = "0"
+        # Filter the test cases by test result
+        for source in self.test_case_sources(measurement.sources):
```

Comment:
maybe the filtering could be incorporated in the loop s of lines 51-55? This might safe an extra loop an potentially performance.

Reply:
There will be typically only be a handful of sources, so iterating through these twice (once to get the test case sources, i.e. Jira, and once to get the test report sources, i.e. JUnit/TestNG) won't be be a problem for performance I think.